maas/resource_maas_network_interface_tag.go

```diff
@@ -0,0 +1,204 @@
+					return nil, err
+				}
+				// Gets the existing interface, thereby ensuring that the node and interface exists.
+				existingInterface, err := client.NetworkInterface.Get(systemId, interfaceId)
+				if err != nil {
+					return nil, err
+				}
+				// Infer the type of the relevant entity from the system ID. Makes calls to MAAS.
+				entityType, err := getMachineOrDeviceTypeFromSystemID(client, systemId)
+				if err != nil {
+					return nil, err
+				}
+				// Set the machine or device in state, set once.
+				if entityType == "machine" {
+					if err := d.Set("machine", existingInterface.SystemID); err != nil {
+						return nil, err
+					}
+				} else {
+					if err := d.Set("device", existingInterface.SystemID); err != nil {
+						return nil, err
+					}
+				}
+				// Set the resource ID
+				d.SetId(fmt.Sprintf("%v:%v", existingInterface.SystemID, existingInterface.ID))
+				return []*schema.ResourceData{d}, nil
+			},
+		},
+		Schema: map[string]*schema.Schema{
+			"device": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ExactlyOneOf: []string{"machine", "device"},
+				Description:  "The identifier (system ID, hostname, or FQDN) of the device with the network interface. Either `machine` or `device` must be provided.",
+			},
+			"interface_id": {
+				Type:        schema.TypeInt,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The network interface ID to tag.",
+			},
+			"machine": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ExactlyOneOf: []string{"machine", "device"},
+				Description:  "The identifier (system ID, hostname, or FQDN) of the machine with the network interface. Either `machine` or `device` must be provided.",
+			},
+			"tags": {
+				Type:        schema.TypeSet,
+				Required:    true,
+				Description: "The tags to assign to the network interface. Tag names should be short and without spaces.",
+				Elem:        &schema.Schema{Type: schema.TypeString},
+			},
+		},
+	}
+}
+
+func resourceNetworkInterfaceTagCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*ClientConfig).Client
+
+	interfaceId := d.Get("interface_id").(int)
+	desiredTags := convertToStringSlice(d.Get("tags").(*schema.Set).List())
+	systemId, err := getMachineOrDeviceSystemID(client, d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	// Get the existing interface
+	existingInterface, err := client.NetworkInterface.Get(systemId, interfaceId)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	// Remove tags that are not in the desired set
+	if existingInterface.Tags == nil {
+		existingInterface.Tags = []string{}
+	}
+	for _, tag := range existingInterface.Tags {
+		if !slices.Contains(desiredTags, tag) {
+			client.NetworkInterface.RemoveTag(systemId, interfaceId, tag)
+		}
+	}
+
+	// Add tags that are in the desired set. AddTag will not add duplicates.
```

Comment:
But still you know which tags you can skip adding. What about adding a similar check to remove in order to avoid unnecessary HTTP calls to MAAS?